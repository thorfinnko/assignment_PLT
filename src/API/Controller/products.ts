import APIBaseClass from '../ApiBase';
import { API_STATUS_CODES } from '../../constants/statuses';

export default class General {
  private static APIBase: APIBaseClass;

  static getAPIBase() {
    if (!General.APIBase) {
      General.APIBase = new APIBaseClass('General.ts');
    }
    return General.APIBase;
  }

  static async fetch(url: string) {
    const response: any = await this.getAPIBase().get(url);
    if (response.status === API_STATUS_CODES.OK) {
      return response;
    }

    return false;
  }
}
