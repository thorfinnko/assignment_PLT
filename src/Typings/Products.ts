export type Product = {
  id: number;
  colour: string;
  name: string;
  price: number;
  img: string;
};
