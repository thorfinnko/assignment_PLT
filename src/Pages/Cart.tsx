import React from 'react';
import { Dimensions, FlatList, StyleSheet, Text, View } from 'react-native';
import { useSelector } from 'react-redux';
import { PriceBottomHeader } from '../Components/PriceBottomHeader';
import { CartCard } from '../Components/CartCard';

function Cart() {
  const { products } = useSelector((state) => state.cart);
  if (products.length > 0) {
    return (
      <View style={{ flex: 1 }}>
        <FlatList
          renderItem={({ item }) => <CartCard item={item} />}
          data={products}
          numColumns={2}
          showsVerticalScrollIndicator={false}
          horizontal={false}
        />
        <View style={{ position: 'absolute', bottom: 0, width: Dimensions.get('screen').width }}>
          <PriceBottomHeader
            price={products.reduce((a, b) => {
              return a + b?.price;
            }, 0) || 0}
          />
        </View>
      </View>
    );
  }

  return (
    <View style={{flex:1, justifyContent:'center', alignItems:'center'}}>
      <Text>No Product Added to cart</Text>
    </View>
  )

}

export default Cart;
