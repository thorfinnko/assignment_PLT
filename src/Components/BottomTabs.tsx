import React from 'react';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import Products from '../Pages/Products';
import Cart from '../Pages/Cart';
import { CartIcon, ProductListIcon } from '../../assets/icons/ProductIcons';

function BottomTabs() {
  const Tab = createBottomTabNavigator();
  return (
    <Tab.Navigator
      initialRouteName="Products"
      screenOptions={({ route }) => ({
        headerStyle: { backgroundColor: '#fdbedb' },
        headerTintColor: '#fff',
        tabBarIcon: () => {
          console.log(route.name, 'route.name')
          let icon;
          if (route.name === 'Products') {
            icon =  <ProductListIcon />
          } else if (route.name === 'Cart') {
            icon = <CartIcon />;
          }
          return icon;
        },
      })}
    >
      <Tab.Screen
        name="Products"
        component={Products}
        options={{
          tabBarLabel: 'PRODUCTS',
          title: 'Products',
        }}
      />
      <Tab.Screen
        name="Cart"
        component={Cart}
        options={{
          tabBarLabel: 'CART',
          title: 'Cart',
        }}
      />
    </Tab.Navigator>
  );
}

export default BottomTabs;
