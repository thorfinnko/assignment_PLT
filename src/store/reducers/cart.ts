import { createSlice } from '@reduxjs/toolkit';
import { Product } from '../../Typings/Products';

type cartReducerInitialState = {
  products: Array<Product>;
};

const initialState: cartReducerInitialState = {
  products: [],
};

const cartSlice = createSlice({
  name: 'cart',
  initialState,
  reducers: {
    addToCart: (state, action) => {
      state.products = [...state.products, action.payload];
    },
    removeFromCart: (state, action) => {
      state.products = state.products.filter((item) => item.id != action.payload.id);
    },
  },
});

export const { addToCart, removeFromCart } = cartSlice.actions;

export default cartSlice.reducer;
