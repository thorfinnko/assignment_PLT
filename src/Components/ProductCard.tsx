import React from 'react';
import { Dimensions, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { Product } from '../Typings/Products';
import FastImage from 'react-native-fast-image';
import { useDispatch } from 'react-redux';
import { addToCart } from '../store/reducers/cart';

export interface ProductCardProps {
  item: Product;
}

export const ProductCard = ({ item }: ProductCardProps) => {
  const dispatch = useDispatch();
  return (
    <View style={styles.cardContainer}>
      <FastImage
        style={{ width: Dimensions.get('screen').width / 2, height: 300 }}
        source={{
          uri: item.img,
          priority: FastImage.priority.normal,
        }}
        resizeMode={FastImage.resizeMode.stretch}
      />
      <Text style={styles.description}>{item.name}</Text>
      <Text style={styles.price}>
        {new Intl.NumberFormat('de-DE', { style: 'currency', currency: 'GBP' }).format(item.price)}
      </Text>
      <Text style={styles.price}>{item.colour}</Text>
      <TouchableOpacity
        activeOpacity={1}
        accessibilityLabel="add to cart"
        style={styles.addToCartButton}
        onPress={() => dispatch(addToCart(item))}
      >
        <Text style={{ fontSize: 12, color: 'white' }}>Add to Cart</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  cardContainer: {
    maxWidth: Dimensions.get('screen').width / 2,
    backgroundColor: 'white',
  },
  description: {
    paddingVertical: 10,
    paddingLeft: 10,
    fontSize: 14,
    color: '#979797',
  },
  price: {
    fontSize: 14,
    color: '#333',
    paddingLeft: 10,
  },
  availableColor: {
    width: 15,
    height: 15,
    marginTop: 6,
    marginLeft: 10,
    borderRadius: 7,
  },
  addToCartButton: {
    position: 'absolute',
    bottom: 8,
    right: 8,
    backgroundColor: '#fdbedb',
    borderRadius: 4,
    padding: 4,
  },
});
