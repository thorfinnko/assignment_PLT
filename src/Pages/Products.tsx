import React, { useEffect } from 'react';
import { FlatList, Text, View } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { fetchProducts } from '../store/reducers/products';
import { ProductCard } from '../Components/ProductCard';
import { Product } from '../Typings/Products';

function Products(): JSX.Element {
  const dispatch = useDispatch();
  const { products, loading, error } = useSelector((state) => state.products);

  useEffect(() => {
    // Dispatch the fetchProducts action
    dispatch(fetchProducts());
  }, [dispatch]); // Include dispatch as a dependency

  if (loading) {
    return (
      <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
        <Text>Loading...</Text>
      </View>
    );
  }

  if (error) {
    return (
      <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
        <Text>Error: {error.message}</Text>
      </View>
    );
  }

  // Render the products once they are loaded
  return (
    <View style={{ flex: 1 }}>
      <FlatList
        renderItem={({ item }) => <ProductCard item={item} />}
        data={products}
        numColumns={2}
        showsVerticalScrollIndicator={false}
        horizontal={false}
      />
    </View>
  );
}

export default Products;
