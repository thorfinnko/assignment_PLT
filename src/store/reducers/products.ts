import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { Product } from '../../Typings/Products';

type productInitialState = {
  loading: boolean;
  isError: boolean;
  products: Array<Product>;
};

const initialState: productInitialState = {
  loading: false,
  isError: false,
  products: [],
};

type payload = Array<Product>;

const productSlice = createSlice({
  name: 'products',
  initialState,
  reducers: {
    fetchProducts: (state: productInitialState) => {
      state.loading = true;
    },
    fetchProductsSuccess: (state: productInitialState, action: PayloadAction<payload>) => {
      state.products = action.payload;
      state.loading = false;
      state.isError = false;
    },
    fetchProductsFailure: (state: productInitialState) => {
      state.loading = false;
      state.isError = true;
    },
  },
});

export const { fetchProducts, fetchProductsSuccess, fetchProductsFailure } = productSlice.actions;

export default productSlice.reducer;
