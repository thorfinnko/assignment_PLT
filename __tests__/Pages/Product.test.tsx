import React from 'react';
import { render, screen, waitFor } from '@testing-library/react-native';
import Cart from '../../src/Pages/Cart';
import { useDispatch, useSelector } from 'react-redux';
import { Provider } from 'react-redux';
import createMockStore from '../../__mocks__/mockStore';
import Products from '../../src/Pages/Products';
import General from '../../src/API/Controller/products';
import { Text, TouchableOpacity, View } from 'react-native';
import { ProductCard, ProductCardProps } from '../../src/Components/ProductCard';
import { addToCart } from '../../src/store/reducers/cart';
import { fetchProducts } from '../../src/store/reducers/products';

const initialState = {
    products: {
        loading: true,
        isError: false,
        products: [],
      },
    cart: {
        products: [],
    },
};

const storeWithData = {
    products: {
        loading: false,
        isError: false,
        products: [
            {
                id: 1,
                colour: "Black",
                name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
                price: 10,
                img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024"
            },
            {
                id: 2,
                colour: "Stone",
                name: "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
                price: 4,
                img: "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024"
            }],
    },
    cart: {
        products: [],
    },
}

const store = createMockStore(initialState);

const customStore = createMockStore(storeWithData)

// jest.mock('../../src/API/Controller/products')

// function MockProductCard(cardProps: ProductCardProps) {
//     const {item} = cardProps
//     return (
//       <View>
//         <Text accessibilityLabel="mock onSelect Input" >
//             {item.name}
//         </Text>
//         <TouchableOpacity
//             accessibilityLabel="mock add to cart"
//             onPress={() => {
//                 store.dispatch(addToCart)
//             }}
//         >
//         </TouchableOpacity>

//       </View>
//     );
// }

// jest.mock('../../src/Components/ProductCard', () => MockProductCard)

describe('test the Products Page', () => {
    it('when there are no products in cart', async () => {
        const {toJSON} = render(
            <Provider store={customStore}>
                <Products />
            </Provider>
        );
        expect(screen).toBeTruthy();
        expect(toJSON()).toMatchSnapshot();
        await waitFor(() => {
          expect(screen.getByText('Loading...')).toBeDefined();
        });
    })

    it('when the products are fetched in List', async () => {
        render(
            <Provider store={customStore}>
                <Products />
            </Provider>
        ) ;

        await waitFor(() => {
            expect(screen.getByText('Product 1')).toBeTruthy();
            expect(screen.getByText('Product 2')).toBeTruthy();
        });
    })
})

