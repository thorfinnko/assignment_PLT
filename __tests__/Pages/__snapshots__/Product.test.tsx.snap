// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the Products Page when the products are fetched in List 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Loading...
  </Text>
</View>
`;

exports[`test the Products Page when there are no products in cart 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Loading...
  </Text>
</View>
`;
